File: Other_Tasks/logic_circuit.cpp
#include <iostream>
#include <memory>

struct Node {
    std::shared_ptr <Node> parent;
    std::shared_ptr <Node> left;
    std::shared_ptr <Node> right;

    virtual ~Node() {};
    virtual bool get_value() const = 0;
};

struct Leaf : public Node {
    private:
        bool value;
    public:
        Leaf(bool val) : value(val) {}
        bool get_value() const override {
            return value;
        }
};

struct And_Node : public Node {
    bool get_value() const override {
        return left->get_value() && right->get_value();
    }
};

struct Or_Node : public Node {
    bool get_value() const override {
        return left->get_value() || right->get_value();
    }
};

template <typename T>
bool evaluate(const T& node) {
    return node->get_value();
};

int main() {
    auto e4 = std::make_shared<And_Node>();
    e4 -> left = std::make_shared<Leaf>(true);
    e4 -> right = std::make_shared<Leaf>(false);

    auto e2 = std::make_shared<Or_Node>();
    e2 -> left = e4;
    e2 -> right = std::make_shared<Leaf>(true);

    auto e5 = std::make_shared<Or_Node>();
    e5 -> left = std::make_shared<Leaf>(false);
    e5 -> right = std::make_shared<Leaf>(false);
    
    auto e6 = std::make_shared<Or_Node>();
    e6 -> left = std::make_shared<Leaf>(true);
    e6 -> right = std::make_shared<Leaf>(false);

    auto e3 = std::make_shared<And_Node>();
    e3 -> left = e5;
    e3 -> right = e6;

    auto e1 = std::make_shared<And_Node>();
    e1 -> left = e2;
    e1 -> right = e3;

    std::cout << "Output: " << evaluate(e1) << std::endl;

    return 0;
}

File: Other_Tasks/logic_circuit.py
from abc import ABC, abstractmethod

class Node(ABC):
    def __init__(self):
        self.parent = None
        self.left = None
        self.right = None
    @abstractmethod
    def get_value(self) -> bool:
        pass

class Leaf(Node):
    def __init__(self, val : bool):
        super().__init__()
        self.value = val
    def get_value(self) -> bool:
        return self.value
    
class And_Node(Node):
    def get_value(self) -> bool:
        return self.left.get_value() and self.right.get_value()
    
class Or_Node(Node):
    def get_value(self) -> bool:
        return self.left.get_value() or self.right.get_value()

def evaluate(node : Node) -> bool:
    return node.get_value()

e4 = And_Node()
e4.left, e4.right = Leaf(True), Leaf(False)

e2 = Or_Node()
e2.left, e2.right = e4, Leaf(True)

e5 = Or_Node()
e5.left, e5.right = Leaf(False), Leaf(False)

e6 = Or_Node()
e6.left, e6.right = Leaf(True), Leaf(False)

e3 = And_Node()
e3.left, e3.right = e5, e6

e1 = And_Node()
e1.left, e1.right = e2, e3

print(f"Result is: {evaluate(e1)}") # # False